[export] Support user input mutation. [1/2]

zhxchen17 opened the pull request:
Summary:
Serialization not implemented yet. Will do in the next diff.

Resolving Github issues:
https://github.com/pytorch/pytorch/issues/112429
https://github.com/pytorch/pytorch/issues/114142

Test Plan:
buck2 run mode/opt caffe2/test:test_export -- -r test_export_
input_mutation

Differential Revision: D51556962




cc @avikchaudhuri @gmagogsfm @tugsbayasgalan @angelayi @suo @ydwu4

zhxchen17 commented:
@pytorchbot rebase

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -559,6 +559,84 @@ def export(
+            if isinstance(arg, torch.Tensor):
+                assert not hasattr(gm_torch_level, node.name)
+                gm_torch_level.register_buffer(node.name, arg)
+                node.replace_all_uses_with(gm_torch_level.graph.get_attr(node.name))
```
Should we copy the metadata here?

zhxchen17 replied:
doesn't seem to break anything if we don't. but I can do that

tugsbayasgalan replied:
Hmm I guess it is because we are still using the correct Fake tensor for each node so it re-populates it fine. In theory, if the meta field has more stuff than example_val, i think we should copy it.

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -559,6 +559,84 @@ def export(
+                gm_torch_level.register_buffer(node.name, arg)
+                node.replace_all_uses_with(gm_torch_level.graph.get_attr(node.name))
+    for node in list(gm_torch_level.graph.nodes):
+        if node.op == "placeholder":
```
Why would there be extra placeholders?

zhxchen17 replied:
because replace_all_uses_with() doesn't remove the replaced nodes.

tugsbayasgalan replied:
I think you can just call graph.erase_node(node_you_want_to_delete) right after replace_all_uses_with ?

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -1721,6 +1722,42 @@ def forward(self, scores, mask):
         optimized_model = torch.compile(exported_model)
         optimized_model(tensor_cpu, mask_cpu)
 
+    def test_export_input_mutation_static_shape(self):
```
This might be too edge case-y: What happens when you have a user input that doesn't require grad but later mutated to require grad? I remember something about buffer is not allowed to have a grad or sth?

zhxchen17 replied:
I just tested with this case and it seems to fail on dynamo even before aot autograd: "a view of a leaf Variable that requires grad is being used in an in-place operation."

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -559,6 +559,84 @@ def export(
+    names = set(user_input_names)
+
+    placeholders = [node for node in gm.graph.nodes if node.op == "placeholder"]
+    start = len(graph_signature.parameters)
```
nit: Can you leave a comment saying that "user inputs are always added in the end"

zhxchen17 committed:
[export] Support user input mutation. [1/2] (#114496)

Summary:

Serialization not implemented yet. Will do in the next diff.

Resolving Github issues:
https://github.com/pytorch/pytorch/issues/112429
https://github.com/pytorch/pytorch/issues/114142

Test Plan: buck2 run mode/opt caffe2/test:test_export -- -r test_export_input_mutation

Differential Revision: D51556962

bdhirsh commented on torch/_export/__init__.py:
```diff
@@ -559,6 +559,88 @@ def export(
+            user_input_names.append(node.name)
+            if isinstance(arg, torch.Tensor):
+                assert not hasattr(gm_torch_level, node.name)
+                gm_torch_level.register_buffer(node.name, arg)
```
Hmmm, what will happen in these two cases?

(1) Model has a forward input that requires_grad=True, and passes that input to a triton kernel

(2) Model has a parameter, and passes it to a triton kernel

zhxchen17 replied:
@bdhirsh sorry I just saw your comment today. I was planning to play around triton kernels today but was distracted. Here is what I tried so far:
1. Passing a forward input with requires_grad=True and mutate it inside forward. TorchDynamo actually errors out before we enter aot_export stage with TorchRuntimeError.
2. Added parameter and mutates it inside forward function. TorchDynamo also errors on that with TorchRuntimeError.

zhxchen17 replied:
Circling back on this:
for case (1) you mentioned I got this error:
```
RuntimeError: Found a graph input that requires gradients, and received a mutation.
This is currently banned in the aot_export workflow. If you need this functionality, please file a github issue.
```
for case (2):
```
RuntimeError: Found a graph input that requires gradients, and received a mutation.
This is currently banned in the aot_export workflow. If you need this functionality, please file a github issue.
```
These are expected from my point of view.

test case I used:
```
    def test_export_input_mutation_static_shape_triton(self):
        from torch.testing._internal.triton_utils import (
            add_kernel,
        )
        import triton

        class MutationModelTriton(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.param = torch.nn.Parameter(torch.randn(12))

            def forward(self, x, y):
                n_elements = y.numel()
                grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']),)
                add_kernel[grid](self.param, x, y, n_elements, BLOCK_SIZE=16)
                return x, y

        inputs = (torch.randn(12, device="cuda"), torch.zeros(12, device="cuda", requires_grad=True))
        model = MutationModelTriton()
        ep = torch.export.export(model, inputs)

```

bdhirsh replied:
> These are expected from my point of view.

Oh are we ok with these being errors? My assumption was that people might end up trying to export models that pass arbitrary tensors into triton kernels (e.g. weights or inputs), but I could be wrong.

zhxchen17 replied:
@bdhirsh Yes we're ok with these being errors. The rationale is, if user saw error like this when exporting a model, it will be very easy for them to change inputs to requires_grad=False without changing modeling or exporting code (most of the cases, people only want forward graph anyway). Also, even for normal aten op we don't allow this in dynamo (my first comment), so I think triton's behavior looks consistent here actually. 

The error message itself subjectively could be improved tho lol, like "hint: triton kernel will assume every input is mutated". I'm not very good at error messaging but I think the error itself looks reasonable enough to me.
